blog/content/post/2022-06-29-xen-debugging-and-development-environment.md

```diff
@@ -0,0 +1,179 @@
+```
+
+I cloned the Xen source and I run the docker container.
+
+```bash
+$ git clone git@github.com:xen-project/xen.git
+$ cd xen
+$ docker run --rm -it -v $PWD:/home/xen -w /home/xen 3mdeb/xen-docker /bin/bash
+```
+Then I built the Xen package:
+
+```
+(docker-container)$ git checkout <version>
+(docker-container)$ ./configure --enable-githttp --enable-systemd
+# there is time now to customize .config
+(docker-container)$ make debball
+```
+
+> Note: `--enable-systemd` requires a `libsystemd-dev` package to be installed
+> in the container.
+
+Build result will be placed in `$XEN_SRC_DIR/dist` as
+`xen-upstream-<version>.deb`. For Debian-based systems, it is easy to install
+it with `dpkg` or `apt`. The package contains all necessary components for
+the host OS along with the Xen kernel.
+
+> Note that the host OS still will require Dom0 Kernel for Xen
+
+To update the pxe-server with new Xen image, install the
+`xen-upstream-<version>.deb` in rootfs which hosts the VMs and copy the Xen
+kernel from `/boot/xen-<version>.gz` to tftpboot/httpboot directory (gunzip the
+kernel first).
+
+After whole this effort, I was able to boot my freshly built Xen on apu2c4 with
+Debian host and Debian guest OS. Now that I have prepared the developing
+procedure I can start narrowing down all the issues.
+
+## Possible use cases
+
+The ability to build Xen with custom changes opens the door to creating
+[custom solutions and products](https://3mdeb.com/hypervisors-development/).
+In addition, it provides an opportunity to workaround well-known problems and
+helps fix them in the mainline in the long term. A adequate example of this
+approach is [adding support for the Trenchboot in the Xen](https://blog.3mdeb.com/2020/2020-10-15-xen-implementation-for-trenchboot/).
+As described in the blog post, our engineers added workaround support in Xen for
+Secure Startup via SKINIT. With the help of Andrew Cooper (Xen maintainer),
+after a few code changes, the
+[patch](https://xenbits.xen.org/gitweb/?p=xen.git;a=commit;h=e4283bf38aae6c2f88cdbdaeef0f005a1a5f6c78)
+was merged into the main.
+
+As a long-time maintainer of PC Engines platforms, 3mdeb sees the opportunity to
+create hardened router products that help maintain the security of the corporate
+network. In addition, custom Xen gives ability to verify various firmware
+features e.g. IOMMU configuration and many others which are verified by Xen
+kernel while booting.
+
+Furthermore, once a custom Xen design has been created, the stability of the
+solution must be ensured. The easiest way to achieve this is to create a CI/CD
+that builds and tests the solution. Our test automation team has historically
+provided Xen tests for our PC Engines platform in every release. This is now
+quite outdated, as the tests were run using Xen 4.8 (the current version of Xen
+at the date of this blog post is 4.17). A description of how to update these
+tests could be a good extension of this blog.
```

Comment:
Laste sentence should more like:

We plan to improve transparent validation infrastructure for [Dasharo](https://dasharo.com/) compatible products soon.